### ISSUE TITLE
Auto-Resume From Latest + Pluggable Warm-Start Loader For User Checkpoints

### ISSUE THREAD

**Issue by mjenrungrot** [mjenrungrot/model-training-framework#10]
# Title
Auto-Resume From Latest + Pluggable Warm-Start Loader For User Checkpoints

## Summary
- Always resume from the latest framework checkpoint if available (full state restore).
- Use resume_from_checkpoint only as a one-time warm start for fine-tuning, delegated to user-provided loader logic for arbitrary formats.
- Keep current explicit resume behavior available and backward-compatible.

## Motivation
- Restarts and preemptions should be seamless and automatic.
- Fine-tuning often uses heterogeneous checkpoints (HF, custom .pt) requiring custom key mapping and selective load.
- Cleanly separate framework-owned fault-tolerant resume from user-owned warm start.

## Current Behavior
- fit() resumes only when resume_from_checkpoint is passed; otherwise starts from scratch.
  - model_training_framework/trainer/core.py:507
- Full resume restores epoch/step/optim/sched/AMP/RNG/dataloader state via _resume_from_checkpoint().
  - model_training_framework/trainer/core.py:1391
- CheckpointManager exposes latest/best but fit() doesn't auto-resume from latest.
  - model_training_framework/trainer/checkpoints.py:644
  - model_training_framework/trainer/checkpoints.py:660
- PreemptionConfig already has resume_from_latest_symlink field (line 625) but it's not used in fit()

## Proposed Behavior
- Precedence inside fit():
  1) If config.preemption.resume_from_latest_symlink and get_latest_checkpoint() returns a path → fully resume via _resume_from_checkpoint(latest).
  2) Else if resume_from_checkpoint is provided and a user warm-start loader is registered → warm-start model weights only (new run at epoch=0/step=0; normal seeding).
  3) Else → start from scratch (seed if configured).
- DDP: Rank 0 performs file I/O; state dicts/results are broadcast with barriers (reuse pattern from _resume_from_checkpoint).

## Public API
- GenericTrainer.set_warm_start_loader(fn: WarmStartLoader) → None
  - Registers a user callback that loads arbitrary checkpoint formats and returns model state dict + strict flag.
- WarmStartLoader protocol (docs-only):
  - def __call__(trainer: GenericTrainer, checkpoint_path: str) -> WarmStartResult
- WarmStartResult (docs-only/dataclass):
  - model_state_dict: dict[str, Any]
  - strict: bool = True

#### Use existing PreemptionConfig
PreemptionConfig already exists at model_training_framework/config/schemas.py:612 and already has:
- `resume_from_latest_symlink: bool = True` (line 625)

No changes needed to PreemptionConfig.

#### Warm-start configuration
Add WarmStartConfig (new) with:
- `loader_class: str | None = None`
- `strict: bool = True`
- Add as `GenericTrainerConfig.warm_start: WarmStartConfig | None = None`

#### Auto-resume behavior
In fit():
- If config.preemption.resume_from_latest_symlink and get_latest_checkpoint() exists → call `_resume_from_checkpoint(latest)`
- If checkpoint is corrupted or incompatible → let it fail with clear error message
- No retry logic - if resume fails, training should stop

#### Clear error messages
- "Failed to resume from checkpoint: {path} - {error}"
- "Checkpoint format incompatible or corrupted"

#### Format detection
Add `CheckpointManager.is_framework_checkpoint(path) -> bool`:
- Load checkpoint with map_location="cpu"
- Check if it has required framework keys: `format_version` and `model_state_dict`
- Return True if it's a framework checkpoint, False otherwise
- Handle exceptions gracefully (corrupted files return False)

#### Validation path in fit()
If resume_from_checkpoint provided:
- If is_framework_checkpoint(path) and no warm-start loader → allow built-in model-only warm start
- If not framework checkpoint and no warm-start loader → raise clear error:
  "Non-framework checkpoint detected; register a warm-start loader via trainer.set_warm_start_loader() or config.warm_start.loader_class"
- Load via user callback when provided; broadcast result

### Type Hints

Add protocol and result dataclass (new file trainer/types.py or near top of core.py):
```python
class WarmStartLoader(Protocol):
    def __call__(self, trainer: GenericTrainer, checkpoint_path: str) -> WarmStartResult: ...

@dataclass
class WarmStartResult:
    model_state_dict: dict[str, Any]
    strict: bool = True
```

Core method types:
- `GenericTrainer.set_warm_start_loader(loader: WarmStartLoader) -> None`
- `_warm_start_from_checkpoint(path: str, loader: WarmStartLoader | None) -> None`

#### Checkpoint format and size
In `_resume_from_checkpoint`:
- Compute size: `Path(path).stat().st_size` (bytes → MB)
- Time the load; log: "Resuming from checkpoint: {path} (size: {size_mb:.2f} MB)"
- Log load time: "Checkpoint loaded in {time:.2f} seconds"

#### Warm-start key mismatches
- Capture load_state_dict return when strict=False
- Log counts + sample keys: missing_keys, unexpected_keys

#### Metrics
Add to PerformanceMonitor or as trainer fields:
- resume_time_sec, checkpoint_size_mb
- Optionally log to experiment logger

#### Auto-resume
- With latest symlink and without; ensure it finds the newest file when symlink missing
- Verify failure on corrupted checkpoint with clear error message

#### Warm-start
- Custom loader with head mismatch; strict=False logs missing/unexpected keys
- No loader + unknown format → raises validation error

#### Precedence rules
- Latest exists + resume_from_checkpoint provided → latest wins, logs ignore message
- No latest + resume_from_checkpoint provided → warm-start path taken

#### DDP robustness
- Only rank 0 reads file; others receive broadcasted dict
- Broadcast failures are properly propagated

#### Migration guide
- If you previously used resume_from_checkpoint for fine-tuning, move to warm-start loader and let auto-resume handle restarts

#### Examples
- Warm-start loader for HuggingFace: strip prefixes, ignore head; return strict=False
- Torchvision model zoo: load model_zoo_url, map keys; show both strict True/False

#### Troubleshooting
- Unknown format errors, mismatch keys, CUDA/CPU mapping advice
- How to disable auto-resume (set resume_from_latest_symlink=False)

## Internal Changes
- Update fit() logic near resume block to check latest → warm-start → scratch.
  - model_training_framework/trainer/core.py:507
- Add GenericTrainer._warm_start_from_checkpoint(path: str, strict: bool) to load weights only via user loader, broadcast in DDP, and keep epoch/step/RNG/optim/sched unchanged.
- Keep _resume_from_checkpoint unchanged except for adding metrics.
  - model_training_framework/trainer/core.py:1391

## Touchpoints
- model_training_framework/trainer/core.py:507
  - Use existing config.preemption.resume_from_latest_symlink
  - Insert precedence logic; integrate warm-start validation+loader
- model_training_framework/trainer/core.py:1391
  - Augment with timing, size logging; keep behavior intact
- model_training_framework/trainer/checkpoints.py:640
  - Add is_framework_checkpoint(path) for format detection (returns bool, not magic string)
- model_training_framework/config/schemas.py
  - No changes to PreemptionConfig (already has resume_from_latest_symlink)
  - Add new WarmStartConfig
  - Attach WarmStartConfig to GenericTrainerConfig

## DDP/Determinism
- Full auto-resume restores RNG and dataloader iteration for deterministic continuation.
- Warm-start intentionally does not restore RNG or dataloader state; it begins a fresh run with seeding.

## Backward Compatibility
- If config.preemption.resume_from_latest_symlink is False, behavior matches today.
- Passing resume_from_checkpoint without a user loader can optionally fall back to a default v1 model-only loader.

## Risks/Edge Cases
- Shape/key mismatches on warm-start → advise strict=False; log clearly.
- resume_from_checkpoint="latest" with no checkpoints → actionable error.
- Checkpoint corruption during auto-resume → fail fast with clear error message
- Large model broadcasts in DDP → consider chunked broadcast for very large models
- Storage accumulation → ensure cleanup policies work correctly with auto-resume

## Acceptance Criteria
- Auto-resume: with latest present and flag enabled, fit() resumes and continues epoch/step with identical batches in DDP.
- Auto-resume failure: corrupted checkpoint causes clear error and stops training
- Warm-start: with no latest and a user loader registered, fit() loads provided weights and starts at epoch=0/step=0 with seeding.
- Format detection: framework vs custom checkpoints are correctly identified via boolean check
- DDP: only rank 0 touches filesystem; broadcasts succeed; barriers ensure consistency.
- Logs indicate path taken (auto-resume vs. warm-start vs. scratch) with timing and size metrics.
- Docs include a complete example warm-start loader.

## Tasks
- Code
  - [ ] Use existing config.preemption.resume_from_latest_symlink in fit()
  - [ ] Add WarmStartConfig to schemas
  - [ ] Add set_warm_start_loader to GenericTrainer
  - [ ] Implement _warm_start_from_checkpoint(...)
  - [ ] Add CheckpointManager.is_framework_checkpoint() for format detection (returns bool)
  - [ ] Modify fit() precedence with validation
  - [ ] Add timing and size metrics to checkpoint operations
- Tests
  - [ ] Single-process: scratch, auto-resume, warm-start
  - [ ] Auto-resume failure on corruption
  - [ ] Format detection tests (framework vs non-framework checkpoints)
  - [ ] DDP: broadcast correctness; non-primary no filesystem I/O
  - [ ] Error handling: missing loader, loader exceptions, strictness mismatch warns
- Docs
  - [ ] API for set_warm_start_loader + WarmStartLoader protocol
  - [ ] Migration guide from old resume_from_checkpoint usage
  - [ ] Quickstart warm-start example
  - [ ] Example loaders for HuggingFace and torchvision
  - [ ] Advanced section explaining precedence and DDP notes
  - [ ] Troubleshooting guide for checkpoint compatibility

## References
- model_training_framework/trainer/core.py:507
- model_training_framework/trainer/core.py:1391
- model_training_framework/trainer/checkpoints.py:644
- model_training_framework/trainer/checkpoints.py:660
- model_training_framework/config/schemas.py:612 (PreemptionConfig)
- model_training_framework/config/schemas.py:625 (resume_from_latest_symlink)

## Labels
- enhancement
